### ISSUE TITLE
In ConEmu, going up in pager glitches

### ISSUE THREAD

**Issue by cmcqueen** [git-for-windows/git#2767]
### Setup

 - Git for Windows 2.28.0-rc2
 - ConEmu 200713

```
$ git --version --build-options
git version 2.28.0.rc2.windows.1
cpu: x86_64
built from commit: 30db709d611267305cf63d5dc704cade83e300ce
sizeof-long: 4
sizeof-size_t: 8
shell-path: /bin/sh
```

 - Windows 10 feature release 2004

```
$ cmd.exe /c ver
Microsoft Windows [Version 10.0.19041.388]
```

```
> type "C:\Program Files\Git\etc\install-options.txt"

Editor Option: Notepad++
Custom Editor Path:
Path Option: Cmd
SSH Option: OpenSSH
Tortoise Option: false
CURL Option: OpenSSL
CRLF Option: CRLFCommitAsIs
Bash Terminal Option: ConHost
Git Pull Behavior Option: Merge
Use Credential Manager: Enabled
Performance Tweaks FSCache: Enabled
Enable Symlinks: Disabled
Enable Pseudo Console Support: Disabled
```

### Details

I did a `git log`. I paged down. All looked fine. I paged up. Instead of paging up as expected, it seemed to show lines glitching on the top row of the console, slowly.

* This only happens in ConEmu, but looks fine in the standard (albeit basic) Windows cmd console.
* This issue doesn't happen with Git for Windows 2.26.2 (in the same version of ConEmu).

Because of the above points, I can't tell if it should be considered a Git for Windows bug or a ConEmu bug.

**Comment by dscho** [comment]
Can you try with the portable Gits provided as [snapshots](https://wingit.blob.core.windows.net/files/index.html)? It is not necessarily obvious, but the snapshot associated with 4267c88d0d8936ae723bba55e0576b1c4f3d4e1a actually refers to v2.26.2, and 30db709d611267305cf63d5dc704cade83e300ce corresponds to v2.28.0-rc2. It would be interesting to know which version is the first one that breaks in ConEmu.

**Comment by ldoron** [comment]
I tested the snapshots. The easiest way I found to reproduce the problems was to run "**less --help**" and "**less -X help**", which sometimes differ a bit in the loss of functionality.

- abf54dc8ea2c1aa9843a073cf86105b8b5801fcb (1 May 2020) was the same as 2.26.2, mostly fine. In both commands, the mouse wheel scrolls the ConEmu window scroll bar rather than sending scroll commands to less.exe (I would expect the latter for "**less --help**").
- 46ad3fc87ad315c4eb557630a375b919c7a81f06 (9 May 2020) began exhibiting problems. **Both commands** no longer show color highlighting of the help text (used for parameters like "N", "pattern", etc.); underlines are used instead. I can scroll up and down by PageDown/PageUp keys, and by line using e/y or j/k keys, but not by DownArrow/UpArrow, which mostly seem to be ignored. Pressing DownArrow/UpArrow causes flashing/repainting in the HELP line; DownArrow works just once if it's the first key I press; and if I hold either one down to repeat, the screen occasionally moves. In "**less --help**", the mouse wheel seems to be treated just like those keys (though, since I can't repeat, I just spin it like crazy). When I quit, the screen (including the ConEmu backscroll buffer) is cleared instead of restoring the previous content. In "**less -X --help**", the mouse wheel scrolls the ConEmu window scroll bar.
- 4df0b592227d5e0a229f2b2173a5696f95ce2f41 (8 Jun 2020) remains the same.
- 3c88881ab77a63b9e11bd35e350e95eef74ee1b0 (16 Jun 2020) changes the symptoms, some for the better, some worse. Again, underlining is used instead of color. In "**less --help**", I can scroll properly with all the keys mentioned above, or with the mouse wheel (_yay!_); the screen/buffer is still cleared when I quit. In "**less -X --help**", I can always scroll down by page or line. If I scroll up by page (b/PageUp) or line (y/k/UpArrow), I get the glitchy writes to the top row and no change in the rest of the screen; if I then repaint (r) the screen, it is corrected and shows the expected view. The mouse wheel scrolls the ConEmu window scroll bar.

Hope this helps!

**Comment by ldoron** [comment]
I also did some quick tests in a standard CMD console. Color highlighting in the first snapshot above (1 May 2020) is replaced by underlines in the subsequent snapshots. In all other respects, as far as I could tell, the snapshots all show the same behavior. All keystroke scrolling is fine with both commands. In "**less --help**", the mouse wheel is completely ignored. When quitting, the screen/buffer is restored properly. In "**less -X --help**", the mouse wheel scrolls the CMD console scroll bar, as expected.

**Comment by dscho** [comment]
I saw some chatter on the Cygwin mailing lists about ConEmu. Hopefully these issues will all be resolved in Cygwin v3.2.0 (and we will inherit those fixes via the MSYS2 runtime).

**Comment by cmcqueen** [comment]
> I saw some chatter on the Cygwin mailing lists about ConEmu.

That's good to know. I couldn't find it, looking in the archives. Which mailing list(s) and dates?

**Comment by dscho** [comment]
I don't recall the dates, but I'm almost certain that I read something about ConEmu on `cygwin-patches`

**Comment by dscho** [comment]
It wasn't `cygwin-patches` after all, it was the commit history. These Cygwin commits talk about ConEmu:

- https://github.com/cygwin/cygwin/commit/d7478090d684ad92773330f1abdff660b537da49
- https://github.com/cygwin/cygwin/commit/50e2a63b04bdd018484605fbb954fd1bd5147fa0
- https://github.com/cygwin/cygwin/commit/0c5aab9c99cf150bc90374c7fb72931e56dc955f

The first two purportedly made it into v3.1.3, the third into v3.1.5. This here report revolves around v2.28.0-rc2 which I believe bundled v3.1.6. Bummer.

**Comment by ldoron** [comment]
Hi, I indicated above (https://github.com/git-for-windows/git/issues/2767#issuecomment-676023256) the exact snapshots where functionality changed in less.exe (1 May to 9 May; 8 Jun to 16 Jun). Can those changes be explained by particular transitions in the MSYS2 & Cygwin versions bundled with the snapshots? What were the versions before and after each of these transitions? Thanks!

**Comment by dscho** [comment]
> Can those changes be explained by particular transitions in the MSYS2 & Cygwin versions bundled with the snapshots?

I would think so.

> What were the versions before and after each of these transitions?

Can you compare the included `etc/package-versions.txt`? They should show the exact version of the `msys2-runtime` package used to generate the snapshot.

**Comment by ldoron** [comment]
Snapshot date|`msys2-runtime` version
-------------|-----------------------
1 May|3.0.7-4
9 May|3.1.4-1
8 Jun|3.1.4-2
16 Jun|3.1.5-1

**Comment by dscho** [comment]
> ```
> 1 May 	3.0.7-4
> 9 May 	3.1.4-1
> ```

That's what I feared. I had waited with the upgrade to v3.1.x for a long time because I did have the impression that the Pseudo Console support was not quite hashed out yet, but it seems that I _still_ jumped the shark.

I updated my Cygwin (including Cygwin's version of `git`) to the newest version and tried the same steps as described at the top of this ticket, and I could reproduce the issue. So it is a Cygwin issue, really.

I even installed [the latest Cygwin snapshot](https://cygwin.com/snapshots/) on top of my Cygwin installation, and the problem _still_ persists. I guess the best bet is to report it there. The closest to a solution is what I found [here](https://cygwin.com/pipermail/cygwin/2020-May/244824.html). The author of Cygwin's Pseudo Console support suggests to use ConEmu's [Cygwin/MSYS connector](https://conemu.github.io/en/CygwinMsysConnector.html). And indeed, if I run this, it works for me:

```cmd
set "PATH=C:\Program Files\Git\usr\bin;%PATH%"
%ConEmuBaseDirShort%\conemu-msys2-64.exe git log
```

Of course, that is _really_ clunky. However, my experiments a là `set GIT_PAGER="%ConEmuBaseDirShort%\conemu-msys2-64.exe" less` did not work:

```cmd
$ set GIT_PAGER="%ConEmuBaseDirShort%\conemu-msys2-64.exe" less
$ git log
Missing filename ("less --help" for help)
```

(Apparently, `conemu-msys2-64.exe` does not preserve the stdin.) I'm not sure where to take it from here.

**Comment by peolsson** [comment]
Thanks for looking into this. I've also tried to find a solution - from both git/Cygwin and ConEmu perspective. However, so far I have found nothing that actually solves the situation. A similar report exists in the ConEmu project (https://github.com/Maximus5/ConEmu/issues/2159). For now the easiest solution is probably to downgrade Git to version 2.27.x.

**Comment by peolssoni3** [comment]
FYI: This has been fixed in latest version of ConEMU: https://github.com/Maximus5/ConEmu/issues/2159
I just tried it, and the pager works as expected again.

**Comment by dscho** [comment]
@peolssoni3 thank you!

Apparently, the fix is supposed to be somewhere in https://github.com/Maximus5/ConEmu/commit/278f3b601f3850dd61b318b63c59e6f484446c22, but I could not spot it. Can you?

**Comment by peolssoni3** [comment]
> @peolssoni3 thank you!
> 
> Apparently, the fix is supposed to be somewhere in [Maximus5/ConEmu@278f3b6](https://github.com/Maximus5/ConEmu/commit/278f3b601f3850dd61b318b63c59e6f484446c22), but I could not spot it. Can you?

I tried, but I didn't find anything obvious. There were so many other formatting changes done at the same time, so it was hard to follow.

**Comment by dscho** [comment]
Maybe @Maximus5 can decipher the commit and point us to the actual fix?

**Comment by Maximus5** [comment]
The whole commit is an actual fix. Scrolling was done in a wrong way in ConEmu.

**Comment by dscho** [comment]
The whole commit also removes dead code, changes indentation and renames variables. Surely I must be misunderstanding your claim that these changes are crucial to fixing some scrolling issue.

**Comment by Maximus5** [comment]
What do you want to understand? The scrolling was done in wrong region, the commit fixes that. Dead code definitely had no effect on that.

**Comment by dscho** [comment]
The question appears because you specifically mentioned that this commit would fix the scrolling issue. Yet the commit is crafted so poorly, under-explaining pretty much everything about it, which prevents other developers from:

- verifying that the fix is correct
- verifying that the fix is complete
- understanding the fix enough to see whether there are similar issues lurking elsewhere in ConEmu's code
- determining whether there is anything that can be done in Cygwin/MSYS2/Git for Windows to avoid similar problems in the future.

Hopefully that makes the problem clearer?
